### ISSUE TITLE
Included file is parsed and its variables resolved on each variable reference

### ISSUE THREAD

**Issue by SkxNMDI5** [serverless/serverless#4466]
# This is a (Bug Report)

## Description
* What went wrong?
I use and include a file (conf.yml) to declares common variables.
One of those is resolved from CloudFormation ```(${cf:otherstack.ArtifactsBucket}```).

This file is referenced in a custom.conf variable.
```
custom:
  conf: ${file(conf.yml)}
```

When deploying, Cloudformation returns a "Rate exceeded" message.
Debugging the code, I can see that serverless is executing the lib/class/Variables.js.getValueFromCf on each reference to a ${self:custom.conf....} variable even if not the one using the CF reference.

For example
     - if my conf file contains 2 variables and only 1 variable ${cf:...}
     - if I reference those 2 vars in my serverless.yml, using the ${self:custom.conf.var} syntax
     - serverless will make 3 calls to the getValueFromCf function (always for the same variable) (1 call when referencing the file + 1 call for each variable ${self:custom.conf.var} resolution)
     - resulting in 3 calls to AWS describe-stacks.
     - In my real case, I use much more references to the conf file and so I reach a "Rate Exceeded" as the calls to describe-stacks are not cached and not serialized.

* What did you expect should have happened?
  1. Only one call to the getValueFromCf function.
  2. No "Rate Exceeded" error.

* What was the config you used?

serverless.yml :
```
service: prototype-sls-cfn-rate

custom:
  conf: ${file(conf.yml)}
  deploymentBucket: ${self:custom.conf.deploymentBucket}
  # following declarations might look stupid but just for showing the issue when referencing conf var
  var1: ${self:custom.conf.appBucket}1
  var2: ${self:custom.conf.appBucket}2
  ....
  # repeat until 'Rate Exceeded' limit is reached on deploy

provider:
  name: aws
  region: eu-west-1
  deploymentBucket: ${self:custom.conf.deploymentBucket}

resources:
  Resources:
    AppBucket:
      Type: 'AWS::S3::Bucket'
      Properties:
        BucketName: ${self:custom.conf.appBucket}
```

conf.yml :
```
deploymentBucket: ${cf:otherstack.ArtifactsBucket}
appBucket: application-used-bucket
```
=> with this configuration getValueFromCf('cf:otherstack.ArtifactsBucket') is called 5 times.

* What stacktrace or error message from your provider did you see?
```
Serverless: Packaging service...
Serverless: Service files not changed. Skipping deployment...

  Serverless Error ---------------------------------------

  Rate exceeded

  Get Support --------------------------------------------
...
```

Similar or dependent issues:
* #3821
* #4136 (same symptom but maybe different cause)

## Additional Data

* ***Serverless Framework Version you're using***: 1.23.0 and 1.24.1
* ***Operating System***: Windows 7
* ***Stack Trace***: see above
* ***Provider Error messages***: see above

## Workaround
Never use ${cf:...} variable resolution in an included file.

## Potential Solutions
* Avoid included file resolution on each referenced variable 
* Use caching as proposed by @stevearoonie in : https://github.com/serverless/serverless/issues/3821#issuecomment-331022279
* Serializing calls won't be enough as too much calls would be done to CloudFormation

**Comment by HyperBrain** [comment]
Seems this is a duplicate of #3821 . Due to the errorneous variable resolution when using AWS calls the `Rate exceeded` error affects every external variable reference (`cf`, `ssm`, etc).

**Comment by SkxNMDI5** [comment]
Hi,
that's right in a way.

But I think this issue makes appear another problem which is the repetitive resolution of included files.

**Comment by HyperBrain** [comment]
Agree, but then the issue should address this problem and reference the other issue as related.

Can you change the subject and point this out in the description. Then it's much easier to get the right understanding 😄

**Comment by HyperBrain** [comment]
@SkxNMDI5 The full fledged solution has now been implemented (see #3821 ) and should solve all variable resolution issues, regardless if CF, SSM, S3 or others.

**Comment by HyperBrain** [comment]
> Use caching as proposed by @stevearoonie in : #3821 (comment)
> Serializing calls won't be enough as too much calls would be done to CloudFormation

The proposed solutions have been implemented as follows:
* variable caching
* request queuing (to restrict execution to 2 at a time)
* request caching (on-demand, activated in the variable system)
